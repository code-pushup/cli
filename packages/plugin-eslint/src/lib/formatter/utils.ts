import { ESLint } from 'eslint';
import { writeFileSync } from 'node:fs';
import { mkdir } from 'node:fs/promises';
import path from 'node:path';
import type { EslintFormat, FormatterConfig } from './types.js';

const eslint = new ESLint();

export function getExtensionForFormat(format: EslintFormat): string {
  const extensionMap: Record<EslintFormat, string> = {
    json: '.json',
    'json-with-metadata': 'json-with-metadata.json',
    html: 'html.html',
    xml: 'xml.xml',
    checkstyle: 'checkstyle.xml',
    junit: 'junit.xml',
    tap: '.tap',
    unix: 'unix.txt',
    stylish: 'stylish.txt',
    compact: 'compact.txt',
    codeframe: 'codeframe.txt',
    table: 'table.txt',
  };

  return extensionMap[format] || '.txt';
}

export async function formatContent(
  results: ESLint.LintResult[],
  format: EslintFormat,
): Promise<string> {
  const formatter = await eslint.loadFormatter(format);
  return formatter.format(results);
}

export function findConfigFromEnv(
  env: NodeJS.ProcessEnv,
): FormatterConfig | null {
  const configString = env['ESLINT_FORMATTER_CONFIG'];

  if (!configString || configString.trim() === '') {
    return null;
  }

  try {
    const parsed = JSON.parse(configString);

    // Validate that parsed result is an object
    if (typeof parsed !== 'object' || parsed == null || Array.isArray(parsed)) {
      console.error('ESLINT_FORMATTER_CONFIG must be a valid JSON object');
      return null;
    }

    return parsed as FormatterConfig;
  } catch (error_) {
    console.error(
      'Error parsing ESLINT_FORMATTER_CONFIG environment variable:',
      (error_ as Error).message,
    );
    return null;
  }
}

export async function handleTerminalOutput(
  format: EslintFormat | undefined,
  results: ESLint.LintResult[],
): Promise<void> {
  if (!format) {
    return;
  }
  const content = await formatContent(results, format);
  // eslint-disable-next-line no-console
  console.log(content);
}

export type PersistConfig = {
  outputDir: string;
  filename: string;
  format: EslintFormat;
  verbose?: boolean;
};

export async function persistEslintReport(
  content: string,
  options: PersistConfig,
): Promise<boolean> {
  const { outputDir, filename, format, verbose = false } = options;
  try {
    await mkdir(path.dirname(outputDir), { recursive: true });
    // eslint-disable-next-line n/no-sync
    writeFileSync(
      path.join(outputDir, `${filename}.${getExtensionForFormat(format)}`),
      content,
    );
    if (verbose) {
      // eslint-disable-next-line no-console
      console.log(`ESLint report (${format}) written to: ${outputDir}`);
    }
    return true;
  } catch (error_) {
    if (verbose) {
      console.error('There was a problem writing the output file:\n%s', error_);
    }
    return false;
  }
}

export async function persistEslintReports(
  formats: EslintFormat[],
  results: ESLint.LintResult[],
  options: { outputDir: string; filename: string; verbose: boolean },
): Promise<boolean> {
  const { outputDir, filename, verbose } = options;

  // eslint-disable-next-line functional/no-loop-statements
  for (const format of formats) {
    const content = await formatContent(results, format);

    const success = await persistEslintReport(content, {
      outputDir,
      filename,
      format,
      verbose,
    });

    if (!success) {
      return false;
    }
  }
  return true;
}
