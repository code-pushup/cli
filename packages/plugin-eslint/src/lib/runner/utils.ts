import type { ESLint } from 'eslint';
import { glob } from 'glob';
import type { PluginArtifactOptions } from '@code-pushup/models';
import {
  executeProcess,
  pluralizeToken,
  readJsonFile,
  ui,
} from '@code-pushup/utils';
import type { LinterOutput } from './types.js';

async function resolveGlobPatterns(patterns: string[]): Promise<string[]> {
  const resolvedPathArrays = await Promise.all(
    patterns.map(async pattern => glob(pattern)),
  );

  const resolvedPaths = resolvedPathArrays.flat();

  return [...new Set(resolvedPaths)].sort();
}

export async function loadArtifacts(
  artifacts: PluginArtifactOptions,
): Promise<LinterOutput[]> {
  if (artifacts.generateArtifactsCommand) {
    const { command, args = [] } =
      typeof artifacts.generateArtifactsCommand === 'string'
        ? { command: artifacts.generateArtifactsCommand }
        : artifacts.generateArtifactsCommand;

    // Log the actual command that was executed
    const commandString =
      typeof artifacts.generateArtifactsCommand === 'string'
        ? artifacts.generateArtifactsCommand
        : `${command} ${args.join(' ')}`;
    await ui().logger.log(`$ ${commandString}`);
    await executeProcess({
      command,
      args,
      ignoreExitCode: true,
    });
  }

  const initialArtifactPaths = Array.isArray(artifacts.artifactsPaths)
    ? artifacts.artifactsPaths
    : [artifacts.artifactsPaths];

  const artifactPaths = await resolveGlobPatterns(initialArtifactPaths);

  ui().logger.log(
    `ESLint plugin resolved ${initialArtifactPaths.length} ${pluralizeToken('pattern', initialArtifactPaths.length)} to ${artifactPaths.length} eslint ${pluralizeToken('report', artifactPaths.length)}`,
  );

  return await Promise.all(
    artifactPaths.map(async artifactPath => {
      // ESLint CLI outputs raw ESLint.LintResult[], but we need LinterOutput format
      const results = await readJsonFile<ESLint.LintResult[]>(artifactPath);
      return {
        results,
        ruleOptionsPerFile: {}, // TODO: This could be enhanced to load actual rule options if needed
      };
    }),
  );
}
