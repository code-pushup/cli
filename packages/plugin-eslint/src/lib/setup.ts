import { ESLint } from 'eslint';
import type { ESLintTarget } from './config';

export function setupESLint(eslintrc: ESLintTarget['eslintrc']) {
  return new ESLint({
    ...(typeof eslintrc === 'string'
      ? { overrideConfigFile: eslintrc }
      : { baseConfig: eslintrc }),
    useEslintrc: false,
    errorOnUnmatchedPattern: false,
  });
}
