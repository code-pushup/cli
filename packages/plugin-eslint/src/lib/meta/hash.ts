import { createHash } from 'node:crypto';
import { slugify } from '@code-pushup/utils';

export function ruleIdToSlug(
  ruleId: string,
  options: unknown[] | undefined,
): string {
  const slug = slugify(ruleId);
  if (!options?.length) {
    return slug;
  }
  return `${slug}-${jsonHash(options)}`;
}

export function jsonHash(data: unknown, bytes = 8): string {
  return createHash('shake256', { outputLength: bytes })
    .update(JSON.stringify(data) || 'null')
    .digest('hex');
}
