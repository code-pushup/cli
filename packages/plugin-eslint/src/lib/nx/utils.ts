import type { ProjectConfiguration } from '@nx/devkit';
import { join } from 'node:path';
import { fileExists, toArray } from '@code-pushup/utils';

export async function findCodePushupEslintrc(
  project: ProjectConfiguration,
): Promise<string | null> {
  const name = 'code-pushup.eslintrc';
  // https://eslint.org/docs/latest/use/configure/configuration-files#configuration-file-formats
  const extensions = ['json', 'js', 'cjs', 'yml', 'yaml'];

  // eslint-disable-next-line functional/no-loop-statements
  for (const ext of extensions) {
    const filename = `./${project.root}/${name}.${ext}`;
    if (await fileExists(join(process.cwd(), filename))) {
      return filename;
    }
  }

  return null;
}

export function getLintFilePatterns(project: ProjectConfiguration): string[] {
  const options = project.targets?.['lint']?.options as
    | { lintFilePatterns?: string | string[] }
    | undefined;
  return options?.lintFilePatterns == null
    ? [`${project.root}/**/*`] // lintFilePatterns defaults to ["{projectRoot}"] - https://github.com/nrwl/nx/pull/20313
    : toArray(options.lintFilePatterns);
}

export function getEslintConfig(
  project: ProjectConfiguration,
): string | undefined {
  const options = project.targets?.['lint']?.options as
    | { eslintConfig?: string }
    | undefined;
  return options?.eslintConfig ?? `./${project.root}/.eslintrc.json`;
}
