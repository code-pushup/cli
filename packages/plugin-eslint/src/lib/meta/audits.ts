import type { Audit } from '@code-pushup/models';
import type { ESLint } from 'eslint';
import { ruleIdToSlug } from './hash';
import { RuleData, listRules } from './rules';

export async function listAudits(
  eslint: ESLint,
  patterns: string | string[],
): Promise<Audit[]> {
  const rules = await listRules(eslint, patterns);
  return rules.map(ruleToAudit);
}

export function ruleToAudit({ ruleId, meta, options }: RuleData): Audit {
  const name = ruleId.split('/').at(-1) ?? ruleId;
  const plugin =
    name === ruleId ? null : ruleId.slice(0, ruleId.lastIndexOf('/'));

  const lines: string[] = [
    `ESLint rule **${name}**${plugin ? `, from _${plugin}_ plugin` : ''}.`,
    ...(options?.length ? ['Custom options:'] : []),
    ...(options?.map(option =>
      ['```json', JSON.stringify(option, null, 2), '```'].join('\n'),
    ) ?? []),
  ];

  return {
    slug: ruleIdToSlug(ruleId, options),
    title: meta.docs?.description ?? name,
    description: lines.join('\n\n'),
    docsUrl: meta.docs?.url,
  };
}
