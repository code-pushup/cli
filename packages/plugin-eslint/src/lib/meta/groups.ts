import type { Rule } from 'eslint';
import type { Group, GroupRef } from '@code-pushup/models';
import { objectToKeys, slugify } from '@code-pushup/utils';
import { ruleIdToSlug } from './hash';
import { type RuleData, parseRuleId } from './rules';

type RuleType = NonNullable<Rule.RuleMetaData['type']>;

// docs on meta.type: https://eslint.org/docs/latest/extend/custom-rules#rule-structure
const typeGroups: Record<RuleType, Omit<Group, 'refs'>> = {
  problem: {
    slug: 'problems',
    title: 'Problems',
    description:
      'Code that either will cause an error or may cause confusing behavior. Developers should consider this a high priority to resolve.',
  },
  suggestion: {
    slug: 'suggestions',
    title: 'Suggestions',
    description:
      "Something that could be done in a better way but no errors will occur if the code isn't changed.",
  },
  layout: {
    slug: 'formatting',
    title: 'Formatting',
    description:
      'Primarily about whitespace, semicolons, commas, and parentheses, all the parts of the program that determine how the code looks rather than how it executes.',
  },
};

export function groupsFromRuleTypes(rules: RuleData[]): Group[] {
  const allTypes = objectToKeys(typeGroups);

  const auditSlugsMap = rules.reduce<Partial<Record<RuleType, string[]>>>(
    (acc, { meta: { type }, ruleId, options }) =>
      type == null
        ? acc
        : {
            ...acc,
            [type]: [...(acc[type] ?? []), ruleIdToSlug(ruleId, options)],
          },
    {},
  );

  return allTypes
    .map(type => ({
      ...typeGroups[type],
      refs:
        auditSlugsMap[type]?.map((slug): GroupRef => ({ slug, weight: 1 })) ??
        [],
    }))
    .filter(group => group.refs.length);
}

export function groupsFromRuleCategories(rules: RuleData[]): Group[] {
  const categoriesMap = rules.reduce<Record<string, Record<string, string[]>>>(
    (acc, { meta: { docs }, ruleId, options }) => {
      // meta.docs.category still used by some popular plugins (e.g. import, react, functional)
      // eslint-disable-next-line deprecation/deprecation
      const category = docs?.category;
      if (!category) {
        return acc;
      }
      const { plugin = '' } = parseRuleId(ruleId);
      return {
        ...acc,
        [plugin]: {
          ...acc[plugin],
          [category]: [
            ...(acc[plugin]?.[category] ?? []),
            ruleIdToSlug(ruleId, options),
          ],
        },
      };
    },
    {},
  );

  const groups = Object.entries(categoriesMap).flatMap(([plugin, categories]) =>
    Object.entries(categories).map(
      ([category, slugs]): Group => ({
        slug: `${slugify(plugin)}-${slugify(category)}`,
        title: `${category} (${plugin})`,
        refs: slugs.map(slug => ({ slug, weight: 1 })),
      }),
    ),
  );

  return [...groups].sort((a, b) => a.slug.localeCompare(b.slug));
}
