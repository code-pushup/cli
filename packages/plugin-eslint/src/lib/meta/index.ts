import type { Audit, Group } from '@code-pushup/models';
import type { ESLintTarget } from '../config';
import { groupsFromRuleCategories, groupsFromRuleTypes } from './groups';
import { listRules } from './rules';
import { ruleToAudit } from './transform';

export async function listAuditsAndGroups(
  targets: ESLintTarget[],
): Promise<{ audits: Audit[]; groups: Group[] }> {
  const rules = await listRules(targets);

  const audits = rules.map(ruleToAudit);

  const groups = [
    ...groupsFromRuleTypes(rules),
    ...groupsFromRuleCategories(rules),
  ];

  return { audits, groups };
}
