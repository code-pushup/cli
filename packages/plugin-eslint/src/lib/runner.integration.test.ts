import { ESLint } from 'eslint';
import { mkdir, rm, writeFile } from 'node:fs/promises';
import os from 'node:os';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { SpyInstance, describe, expect, it } from 'vitest';
import type { AuditOutput, AuditOutputs, Issue } from '@code-pushup/models';
import { osAgnosticAuditOutputs } from '@code-pushup/test-utils';
import { readJsonFile } from '@code-pushup/utils';
import { listAuditsAndGroups } from './meta';
import {
  ESLINTRC_PATH,
  RUNNER_OUTPUT_PATH,
  createRunnerConfig,
  executeRunner,
} from './runner';
import { setupESLint } from './setup';

describe('executeRunner', () => {
  let cwdSpy: SpyInstance;
  let platformSpy: SpyInstance;

  const createArgv = async (eslintrc: string) => {
    const patterns = ['src/**/*.js', 'src/**/*.jsx'];
    const eslint = setupESLint(eslintrc);
    const { audits } = await listAuditsAndGroups(eslint, patterns);
    const runnerConfig = createRunnerConfig(
      'bin.js',
      audits,
      eslintrc,
      patterns,
    );
    return [runnerConfig.command, ...(runnerConfig.args ?? [])];
  };

  const appDir = join(
    fileURLToPath(dirname(import.meta.url)),
    '..',
    '..',
    'mocks',
    'fixtures',
    'todos-app',
  );

  beforeAll(async () => {
    cwdSpy = vi.spyOn(process, 'cwd').mockReturnValue(appDir);
    // Windows does not require additional quotation marks for globs
    platformSpy = vi.spyOn(os, 'platform').mockReturnValue('win32');

    const config: ESLint.ConfigData = {
      extends: '@code-pushup',
    };
    await mkdir(dirname(ESLINTRC_PATH), { recursive: true });
    await writeFile(ESLINTRC_PATH, JSON.stringify(config));
  });

  afterAll(async () => {
    cwdSpy.mockRestore();
    platformSpy.mockRestore();

    await rm(ESLINTRC_PATH, { force: true });
  });

  it('should execute ESLint and create audit results for React application', async () => {
    const argv = await createArgv('.eslintrc.js');

    await executeRunner(argv);

    const json = await readJsonFile<AuditOutputs>(RUNNER_OUTPUT_PATH);
    expect(osAgnosticAuditOutputs(json)).toMatchSnapshot();
  });

  it('should execute runner with inline config using @code-pushup/eslint-config', async () => {
    const argv = await createArgv(ESLINTRC_PATH);

    await executeRunner(argv);

    const json = await readJsonFile<AuditOutput[]>(RUNNER_OUTPUT_PATH);
    // expect warnings from unicorn/filename-case rule from default config
    expect(json).toContainEqual(
      expect.objectContaining({
        slug: 'unicorn-filename-case',
        displayValue: '5 warnings',
        details: {
          issues: expect.arrayContaining([
            {
              severity: 'warning',
              message:
                'Filename is not in kebab case. Rename it to `use-todos.js`.',
              source: expect.objectContaining({
                file: join(appDir, 'src', 'hooks', 'useTodos.js'),
              } satisfies Partial<Issue['source']>),
            } satisfies Issue,
          ]),
        },
      } satisfies Partial<AuditOutput>),
    );
  }, 7000);
});
