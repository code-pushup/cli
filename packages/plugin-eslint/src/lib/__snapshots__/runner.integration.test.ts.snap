// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`executeRunner > should execute ESLint and create audit results for React application 1`] = `
[
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "no-cond-assign",
    "value": 0,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "no-const-assign",
    "value": 0,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "no-debugger",
    "value": 0,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "no-invalid-regexp",
    "value": 0,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "no-undef",
    "value": 0,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "no-unreachable-loop",
    "value": 0,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "no-unsafe-negation",
    "value": 0,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "no-unsafe-optional-chaining",
    "value": 0,
  },
  {
    "details": {
      "issues": [
        {
          "message": "'loading' is assigned a value but never used.",
          "severity": "warning",
          "source": {
            "file": "<CWD>/src/App.jsx",
            "position": {
              "endColumn": 18,
              "endLine": 8,
              "startColumn": 11,
              "startLine": 8,
            },
          },
        },
      ],
    },
    "displayValue": "1 warning",
    "score": 0,
    "slug": "no-unused-vars",
    "value": 1,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "use-isnan",
    "value": 0,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "valid-typeof",
    "value": 0,
  },
  {
    "details": {
      "issues": [
        {
          "message": "Unexpected block statement surrounding arrow body; move the returned value immediately after the \`=>\`.",
          "severity": "warning",
          "source": {
            "file": "<CWD>/src/components/TodoFilter.jsx",
            "position": {
              "endColumn": 2,
              "endLine": 25,
              "startColumn": 29,
              "startLine": 3,
            },
          },
        },
      ],
    },
    "displayValue": "1 warning",
    "score": 0,
    "slug": "arrow-body-style",
    "value": 1,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "camelcase",
    "value": 0,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "curly",
    "value": 0,
  },
  {
    "details": {
      "issues": [
        {
          "message": "Expected '===' and instead saw '=='.",
          "severity": "warning",
          "source": {
            "file": "<CWD>/src/hooks/useTodos.js",
            "position": {
              "endColumn": 43,
              "endLine": 41,
              "startColumn": 41,
              "startLine": 41,
            },
          },
        },
      ],
    },
    "displayValue": "1 warning",
    "score": 0,
    "slug": "eqeqeq",
    "value": 1,
  },
  {
    "details": {
      "issues": [
        {
          "message": "Arrow function has too many lines (71). Maximum allowed is 50.",
          "severity": "warning",
          "source": {
            "file": "<CWD>/src/hooks/useTodos.js",
            "position": {
              "endColumn": 2,
              "endLine": 73,
              "startColumn": 25,
              "startLine": 3,
            },
          },
        },
      ],
    },
    "displayValue": "1 warning",
    "score": 0,
    "slug": "max-lines-per-function",
    "value": 1,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "max-lines",
    "value": 0,
  },
  {
    "details": {
      "issues": [
        {
          "message": "'data' is already declared in the upper scope on line 5 column 10.",
          "severity": "warning",
          "source": {
            "file": "<CWD>/src/hooks/useTodos.js",
            "position": {
              "endColumn": 17,
              "endLine": 11,
              "startColumn": 13,
              "startLine": 11,
            },
          },
        },
        {
          "message": "'data' is already declared in the upper scope on line 5 column 10.",
          "severity": "warning",
          "source": {
            "file": "<CWD>/src/hooks/useTodos.js",
            "position": {
              "endColumn": 21,
              "endLine": 29,
              "startColumn": 17,
              "startLine": 29,
            },
          },
        },
        {
          "message": "'data' is already declared in the upper scope on line 5 column 10.",
          "severity": "warning",
          "source": {
            "file": "<CWD>/src/hooks/useTodos.js",
            "position": {
              "endColumn": 17,
              "endLine": 41,
              "startColumn": 13,
              "startLine": 41,
            },
          },
        },
      ],
    },
    "displayValue": "3 warnings",
    "score": 0,
    "slug": "no-shadow",
    "value": 3,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "no-var",
    "value": 0,
  },
  {
    "details": {
      "issues": [
        {
          "message": "Expected property shorthand.",
          "severity": "warning",
          "source": {
            "file": "<CWD>/src/hooks/useTodos.js",
            "position": {
              "endColumn": 19,
              "endLine": 19,
              "startColumn": 7,
              "startLine": 19,
            },
          },
        },
        {
          "message": "Expected property shorthand.",
          "severity": "warning",
          "source": {
            "file": "<CWD>/src/hooks/useTodos.js",
            "position": {
              "endColumn": 19,
              "endLine": 32,
              "startColumn": 13,
              "startLine": 32,
            },
          },
        },
        {
          "message": "Expected property shorthand.",
          "severity": "warning",
          "source": {
            "file": "<CWD>/src/hooks/useTodos.js",
            "position": {
              "endColumn": 25,
              "endLine": 33,
              "startColumn": 13,
              "startLine": 33,
            },
          },
        },
      ],
    },
    "displayValue": "3 warnings",
    "score": 0,
    "slug": "object-shorthand",
    "value": 3,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "prefer-arrow-callback",
    "value": 0,
  },
  {
    "details": {
      "issues": [
        {
          "message": "'root' is never reassigned. Use 'const' instead.",
          "severity": "warning",
          "source": {
            "file": "<CWD>/src/index.jsx",
            "position": {
              "endColumn": 9,
              "endLine": 5,
              "startColumn": 5,
              "startLine": 5,
            },
          },
        },
      ],
    },
    "displayValue": "1 warning",
    "score": 0,
    "slug": "prefer-const",
    "value": 1,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "prefer-object-spread",
    "value": 0,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "yoda",
    "value": 0,
  },
  {
    "details": {
      "issues": [
        {
          "message": "Missing \\"key\\" prop for element in iterator",
          "severity": "warning",
          "source": {
            "file": "<CWD>/src/components/TodoList.jsx",
            "position": {
              "endColumn": 12,
              "endLine": 28,
              "startColumn": 7,
              "startLine": 7,
            },
          },
        },
      ],
    },
    "displayValue": "1 warning",
    "score": 0,
    "slug": "react-jsx-key",
    "value": 1,
  },
  {
    "details": {
      "issues": [
        {
          "message": "'onCreate' is missing in props validation",
          "severity": "warning",
          "source": {
            "file": "<CWD>/src/components/CreateTodo.jsx",
            "position": {
              "endColumn": 23,
              "endLine": 15,
              "startColumn": 15,
              "startLine": 15,
            },
          },
        },
        {
          "message": "'setQuery' is missing in props validation",
          "severity": "warning",
          "source": {
            "file": "<CWD>/src/components/TodoFilter.jsx",
            "position": {
              "endColumn": 25,
              "endLine": 10,
              "startColumn": 17,
              "startLine": 10,
            },
          },
        },
        {
          "message": "'setHideComplete' is missing in props validation",
          "severity": "warning",
          "source": {
            "file": "<CWD>/src/components/TodoFilter.jsx",
            "position": {
              "endColumn": 34,
              "endLine": 18,
              "startColumn": 19,
              "startLine": 18,
            },
          },
        },
        {
          "message": "'todos' is missing in props validation",
          "severity": "warning",
          "source": {
            "file": "<CWD>/src/components/TodoList.jsx",
            "position": {
              "endColumn": 17,
              "endLine": 6,
              "startColumn": 12,
              "startLine": 6,
            },
          },
        },
        {
          "message": "'todos.map' is missing in props validation",
          "severity": "warning",
          "source": {
            "file": "<CWD>/src/components/TodoList.jsx",
            "position": {
              "endColumn": 21,
              "endLine": 6,
              "startColumn": 18,
              "startLine": 6,
            },
          },
        },
        {
          "message": "'onEdit' is missing in props validation",
          "severity": "warning",
          "source": {
            "file": "<CWD>/src/components/TodoList.jsx",
            "position": {
              "endColumn": 27,
              "endLine": 13,
              "startColumn": 21,
              "startLine": 13,
            },
          },
        },
      ],
    },
    "displayValue": "6 warnings",
    "score": 0,
    "slug": "react-prop-types",
    "value": 6,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "react-react-in-jsx-scope",
    "value": 0,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "react-hooks-rules-of-hooks",
    "value": 0,
  },
  {
    "details": {
      "issues": [
        {
          "message": "React Hook useCallback does nothing when called with only one argument. Did you forget to pass an array of dependencies?",
          "severity": "warning",
          "source": {
            "file": "<CWD>/src/hooks/useTodos.js",
            "position": {
              "endColumn": 31,
              "endLine": 17,
              "startColumn": 20,
              "startLine": 17,
            },
          },
        },
        {
          "message": "React Hook useCallback does nothing when called with only one argument. Did you forget to pass an array of dependencies?",
          "severity": "warning",
          "source": {
            "file": "<CWD>/src/hooks/useTodos.js",
            "position": {
              "endColumn": 29,
              "endLine": 40,
              "startColumn": 18,
              "startLine": 40,
            },
          },
        },
      ],
    },
    "displayValue": "2 warnings",
    "score": 0,
    "slug": "react-hooks-exhaustive-deps",
    "value": 2,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "react-display-name",
    "value": 0,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "react-jsx-no-comment-textnodes",
    "value": 0,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "react-jsx-no-duplicate-props",
    "value": 0,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "react-jsx-no-target-blank",
    "value": 0,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "react-jsx-no-undef",
    "value": 0,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "react-jsx-uses-react",
    "value": 0,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "react-jsx-uses-vars",
    "value": 0,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "react-no-children-prop",
    "value": 0,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "react-no-danger-with-children",
    "value": 0,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "react-no-deprecated",
    "value": 0,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "react-no-direct-mutation-state",
    "value": 0,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "react-no-find-dom-node",
    "value": 0,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "react-no-is-mounted",
    "value": 0,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "react-no-render-return-value",
    "value": 0,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "react-no-string-refs",
    "value": 0,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "react-no-unescaped-entities",
    "value": 0,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "react-no-unknown-property",
    "value": 0,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "passed",
    "score": 1,
    "slug": "react-require-render-return",
    "value": 0,
  },
]
`;
